Automattic/wp-calypso
# Fix empty-cart redirect when coming from Jetpack Cloud pricing page

Users who arrived at Checkout coming from the Jetpack Cloud pricing page (`cloud.jetpack.com/pricing/:site`), aren't redirected back to the Pricing page if they remove all items from their cart. The current behavior is that Checkout redirects users to `wordpress.com/plans/:site`. 

This PR fixes it so that when coming from Jetpack Cloud pricing page, after removing all items from the cart, you will be redirected back to the Jetpack Cloud pricing page.

Additionally, since we are redirecting the user to a location based on a URL in the query string, this creates an [open redirect](https://cwe.mitre.org/data/definitions/601.html), so I created a `useValidCheckoutBackUrl` hook that only allows url's to the Jetpack Cloud pricing page or the site itself.

#### Changes proposed in this Pull Request
- I Created a `useValidCheckoutBackUrl` hook that retrieves the `checkoutBackUrl` query arg and validates that the hostname is either cloud.jetpack.com or the site hostname.
- In the `useRemoveFromCartAndRedirect` hook, I utilize the `useValidCheckoutBackUrl` hook, and if it's valid, we redirect to it.
- I converted the MasterbarCheckout component to a react functional component so we can easily use the new `useValidCheckoutBackUrl` hook in this component also.

#### Testing instructions

- Checkout and run this PR in calypso blue (`yarn start`).
- Go to `https://cloud.jetpack.com/pricing/<some connected Jetpack site-slug that you own>`
- On the pricing page, select any Jetpack product.
- On the checkout page, in the url, replace `https://wordpress.com` with `calypso.localhost:3000` and press enter to reload the page in the development environment.
- Remove the product from the cart by clicking the `trash` icon, then click the "Continue" button.
- Verify you are redirected back to `https://cloud/jetpack.com/pricing/:site`

Asana task: 1164141197617539-as-1200486811875104
